src/box/memtx_tx.c

```diff
@@ -1190,6 +1211,158 @@ memtx_tx_track_read(struct txn *txn, struct space *space, struct tuple *tuple)
 	return 0;
 }
 
+struct memtx_entry memtx_key_dup(struct txn *txn, struct memtx_entry entry)
+{
+	const char* key = entry.key;
+	if (key == NULL)
+		return entry;
+
+	if (key == &min_key || key == &max_key)
+		return entry;
+
+	assert(mp_typeof(*key) == MP_ARRAY);
+	const char *end = key;
+	mp_next(&end);
+	entry.key = region_alloc(&txn->region, end - key);
+	if (entry.key == NULL) {
+		diag_set(OutOfMemory, end - key, "region_alloc", "key");
+		return entry;
+	}
+	memcpy((void*)entry.key, key, end - key);
+	return entry;
+}
+
+struct memtx_read_interval *
+memtx_read_interval_new(struct txn *tx, struct index *index,
+		     struct memtx_entry left, bool left_belongs,
+		     struct memtx_entry right, bool right_belongs)
+{
+	assert(left.key != NULL);
+	assert(right.key != NULL);
+
+	struct memtx_read_interval *interval;
+	interval = mempool_alloc(&txm.read_interval_mempool);
+	if (interval == NULL) {
+		diag_set(OutOfMemory, sizeof(*interval),
+			 "mempool", "struct memtx_read_interval");
+		return NULL;
+	}
+	interval->tx = tx;
+	index_ref(index);
+	interval->index = index;
+
+	interval->left = memtx_key_dup(tx, left);
+	interval->left_belongs = left_belongs;
+
+	interval->right = memtx_key_dup(tx, right);
+	interval->right_belongs = right_belongs;
+
+	interval->subtree_last = NULL;
+
+	return interval;
+}
+
+void
+memtx_read_interval_delete(struct memtx_read_interval *interval)
+{
+	index_unref(interval->index);
+	mempool_free(&txm.read_interval_mempool, interval);
+}
+
+int
+memtx_tx_track(struct txn *txn, struct index *index,
+	       struct memtx_entry left, bool left_belongs,
+	       struct memtx_entry right, bool right_belongs)
+{
+	assert(index != NULL);
+
+	if (left.key == NULL)
+		left.key = &min_key;
+	if (right.key == NULL)
+		right.key = &max_key;
+
+	if (index->def->space_id >= BOX_SYSTEM_ID_MIN
```

Comment:
Why here you are ignore system spaces? Below is a simple test case with SIGSEGV! It's fails on master also, so it's not a problem of your patch, but i think it's not good to fail with SIGSEGV here. Maybe @alyapunov or @Korablev77  know is it's valid test case? Can we access system spaces in a transaction?

#! /home/evgeny/TARANTOOL/MAS/tarantool/src/tarantool

os.execute("rm -rf *.snap *.xlog")

box.cfg{
    memtx_use_mvcc_engine  = true,
}

txn_proxy = require('txn_proxy')

tx1 = txn_proxy.new()
tx2 = txn_proxy.new()

tx1:begin()
tx2:begin()

tx1('box.space._user:select{}')
tx2('box.space._user:select{}')

tx1("box.schema.user.create('internal1')")
tx2("box.schema.user.create('internal2')")

tx1:commit()
tx2:commit()